### PR TITLE
Project auth - by Olof & Bäm

### PR DESCRIPTION
**Pull request by luuf81**
**Link to my deployed project:**

Frontend: https://project-auth-bamolof.netlify.app/
Backend: https://project-auth-olof.herokuapp.com/

**My reflections on how this week's project turned out:**

It was fun to finally create a app that brings everything together - backend, frontend, authentication. Will be great lessons for final project! 

Things I'd like to have clarified or explained in more detail:

Some things could probably be further reinforced by looking at the recorded videos/lectures a few more times :)

// Olof & Bäm

### CODE REVIEW

File: frontend/src/components/LoginForm.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+// const SIGNUP_URL = "http://localhost:8080/users";
+const SIGNUP_URL = "https://project-auth-olof.herokuapp.com/users";
+// const LOGIN_URL = "http://localhost:8080/sessions";
+const LOGIN_URL = "https://project-auth-olof.herokuapp.com/sessions";
+
+export const LoginForm = () => {
+  const [name, setName] = useState('')
+  const [password, setPassword] = useState('')
+  
+  const dispatch = useDispatch()
```
Comment (Bluepoisson) [inline frontend/src/components/LoginForm.js:17]: Semicolons to be or not to be? Try to be consistent in the usage of them.

File: frontend/src/components/LoginForm.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  const statusMessage = useSelector((store) => store.user.login.statusMessage)
+  const loggedIn = useSelector((store) => store.user.login.loggedIn)
+
+  const handleLoginSuccess = (loginResponse) => {
+    dispatch(user.actions.setAccessToken({ accessToken: loginResponse.accessToken }));
+    dispatch(user.actions.setUserId({ userId: loginResponse.userId }));
+    dispatch(user.actions.setStatusMessage({ statusMessage: "Login Success" }));
```
Comment (Bluepoisson) [inline frontend/src/components/LoginForm.js:24]: This message is shown when we have logged out. Perhaps change it's scope?

File: frontend/src/components/Profile.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  const userId = useSelector((store) => store.user.login.userId)
+  const loggedIn = useSelector((store) => store.user.login.loggedIn)
+
+  const logout = () => {
+    dispatch(user.actions.logout());
+    dispatch(user.actions.toggleLoggedState(false));
+  }
+
+  if (!loggedIn) {
+    return <div>Please log in</div>
```
Comment (Bluepoisson) [inline frontend/src/components/Profile.js:19]: Where is this div shown?

File: frontend/src/styling/form.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+`
+
+export const ProfileSection = styled.section`
+width: 100%;
+display: flex;
+flex-direction: column;
+justify-content: center;
+box-sizing: border-box;
+`
+
```
Comment (Bluepoisson) [inline frontend/src/styling/form.js:80]: Remove blank spaces :)

File: frontend/src/components/Notes.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
+import React, { useState, useEffect } from 'react'
+import { useSelector } from 'react-redux'
+import { NotesContainer } from './NotesContainer'
+import { Button, ProfileSection } from '../styling/form'
```
Comment (Bluepoisson) [inline frontend/src/components/Notes.js:4]: Great with styling components, keeps the code clean.

**Review by Bluepoisson — APPROVED** [review]
Great job Bäm and Olof! You've created a stylish frontend with a complex backend enabling user to post and save notes. You successfully connected them 🥇  We also appreciated the wink to the happy thoughts project, reusing the input component 😄 Approved!

File: backend/server.js [hunk 6]
```diff
@@ -2,11 +2,72 @@ import express from 'express'
...
+const User = mongoose.model('User', userSchema);
+
+const Note = mongoose.model('Note', {
+  description: {
+    type: String,
+    required: true
+  },
+  user: {
+  type: mongoose.Schema.Types.ObjectId,
+  ref: 'User'
```
Comment (KarolCislowski) [inline backend/server.js:57]: indentation 😄

File: backend/server.js [hunk 7]
```diff
@@ -20,9 +81,73 @@ app.use(bodyParser.json())
...
+//authenticated endpoint to post notes
+app.post('/notes', authenticateUser)
+app.post('/notes', async (req, res) => {
+  try {
+    const description = req.body.description
+    const user = req.user
+    const note = await new Note({ description, user }).save()
+    res.json({description, user})
+  } catch (err) {
+    res.status(404).json({ error: 'User not found' });
```
[the remark this answers is not in the document]
Comment (KarolCislowski) [reply backend/server.js:135]: this endpoint is adding data so status should be `400` and message should be something like `couldn't create new note`.

File: frontend/src/reducers/user.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    },
+    toggleLoggedState: (state, action) => {
+      state.login.loggedIn = action.payload
+    },
+    logout: (state, action) => {
+      state.login.userId = 0
+      state.login.accessToken = null
+    },
+  },
+})
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:37]: You need to store your user data in the local storage, to prevent logging out after refreshing the page 😺

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
This is an awesome project! It is very ambitious 🌈 

Smart that you reused idea from the happy thoughts! And this picture on the logging page is so nice! 

To get your black belts you need to fix the refresh page=loggout issue! Good luck!

**Review by KarolCislowski — APPROVED** [review]
And now it is a pure magic 🦄 🌈

**Pull request closed without merging**